### PR TITLE
Coordinate all request timeout handling through a min heap on the root channel

### PR DESCRIPTION
**Pull request by jcorbin**
r @Raynos @kriskowal 

Not yet passing tests, early share; see especially the comment block at head of ttl_heap.js

### CODE REVIEW

**Comment by Raynos** [comment]
I approve; lot's of great simplifications.

File: node/ttl_heap.js [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+        var now = self.timers.now();
+        self.onTimeout(now);
+    }
+};
+
+TTLHeap.prototype.onTimeout = function onTimeout(now) {
+    var self = this;
+
+    self.timer = null;
+    self.drainExpired(now);
```
Comment (kriskowal) [inline node/ttl_heap.js]: Unnecessary O(n) operation if you reframe ttl absolute instead of relative time.

File: node/ttl_heap.js [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+    self.lastTime = self.timers.now();
+    self.timer = null;
+}
+
+TTLHeap.prototype.clear = function clear() {
+    var self = this;
+
+    self.timers.clearTimeout(self.timer);
+    self.items = [];
+    self.ttls = [];
```
Comment (kriskowal) [inline node/ttl_heap.js]: .length = 0 is cheap and eliminates one piece of garbage

**Comment by jcorbin** [comment]
Now ready for review, whilst I write a unit test for the time_heap module

**Comment by Raynos** [comment]
Lets add that 10 second sanity check.

Otherwise lgtm.

**Comment by jcorbin** [comment]
Update: had tests passing last night, now tests don't pass this morning:
- hard to repo flaps local
- however those seemed to stop after an npm install that upgraded debug-logtron
- now I have two actual failures in the new rate limiter test by @ShanniLi , working on those

**Comment by jcorbin** [comment]
Oh also, I switched over to using test/lib/test_search to abstract/aggregate way all the ~70k TimeHeap assertions into one "it worked" pass, or the first failing case if any.

**Comment by Raynos** [comment]
lgtm. ship this.

**Comment by Raynos** [comment]
I noted a memory leak in the benchmark.

All three (client, relay, bench) were sitting at 2GB of memory locally.

**Comment by jcorbin** [comment]
memory leak fixed, landing once travis greens

**Comment by kriskowal** [comment]
:clap: :clap: :clap:

**Pull request merged**